File: Data Structures & Algorithms/valid-binary-search-tree/submission-5.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isValidBST(root) {
        return this.isValidBST_DFS(root, -Infinity, Infinity);
    }

    isValidBST_DFS(root, min, max) {
        if (!root) return true;

        return root.val > min && root.val < max && this.isValidBST_DFS(root.left, min, root.val)
            && this.isValidBST_DFS(root.right, root.val, max);
    }
}

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-5.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        int max = Math.max(p.val, q.val);
        int min = Math.min(p.val, q.val);
        while (root != null) {
            if (root.val <= max && root.val >= min) return root;
            if (root.val > max) root = root.left;
            if (root.val < min) root = root.right;
        }

        return null;
    }
}
